CMakeLists.txt

```diff
@@ -535,6 +539,61 @@ if(USE_NGTCP2)
   set(USE_NGHTTP3 ON)
   include_directories(${NGHTTP3_INCLUDE_DIRS})
   list(APPEND CURL_LIBS ${NGHTTP3_LIBRARIES})
+
+  if(WIN32)
+    cmake_push_check_state()
+    set(CMAKE_REQUIRED_LIBRARIES "${CURL_LIBS}")
+    check_symbol_exists(nghttp3_version "nghttp3/nghttp3.h" HAVE_NGHTTP3_DYNAMICLIB)
+    if(NOT HAVE_NGHTTP3_DYNAMICLIB)
+      set(CMAKE_REQUIRED_DEFINITIONS "-DNGHTTP3_STATICLIB=1")
+      check_symbol_exists(nghttp3_version "nghttp3/nghttp3.h" HAVE_NGHTTP3_STATICLIB)
+      if(HAVE_NGHTTP3_STATICLIB)
+        add_definitions(-DNGHTTP3_STATICLIB=1)
+      endif()
+    endif()
+
+    check_symbol_exists(ngtcp2_version "ngtcp2/ngtcp2.h" HAVE_NGTCP2_DYNAMICLIB)
+    if(NOT HAVE_NGTCP2_DYNAMICLIB)
+      set(CMAKE_REQUIRED_DEFINITIONS "-DNGTCP2_STATICLIB=1")
+      check_symbol_exists(ngtcp2_version "ngtcp2/ngtcp2.h" HAVE_NGTCP2_STATICLIB)
+      if(HAVE_NGTCP2_STATICLIB)
+        add_definitions(-DNGTCP2_STATICLIB=1)
+      endif()
+    endif()
+    cmake_pop_check_state()
+  endif()
+endif()
+
+option(USE_NGHTTP2 "Use Nghttp2 library" OFF)
+if(USE_NGHTTP2)
+  find_package(NGHTTP2 REQUIRED)
+  include_directories(${NGHTTP2_INCLUDE_DIRS})
+  list(APPEND CURL_LIBS ${NGHTTP2_LIBRARIES})
+  if(WIN32)
+    cmake_push_check_state()
+    set(CMAKE_REQUIRED_LIBRARIES "${CURL_LIBS}")
+    check_type_size(ssize_t SIZEOF_SSIZE_T)
+    if(NOT HAVE_SIZEOF_SSIZE_T)
+      check_type_size("long" SIZEOF_LONG)
+      check_type_size("__int64" SIZEOF___INT64)
+      if(SIZEOF_LONG EQUAL SIZEOF_SIZE_T)
+        set(CMAKE_REQUIRED_DEFINITIONS "-Dssize_t=long")
+      elseif(SIZEOF___INT64 EQUAL SIZEOF_SIZE_T)
+        set(CMAKE_REQUIRED_DEFINITIONS "-Dssize_t=__int64")
+      else()
+        set(CMAKE_REQUIRED_DEFINITIONS "-Dssize_t=long long")
+      endif()
```

Comment:
This type defines seem unrelated to a static/dynamic build. Why do you need those?

Reply:
nghttp2 use `ssize_t` but not define it.

Reply:
The failed jobs seem not a problem of this PR, could you please help me to find out what's the problem?

Reply:
> nghttp2 use `ssize_t` but not define it.

We've built curl with nghttp2 for many years already without having to add any extra defines. Why does the need start now? And if we actually do need them, they are still unrelated to the static builds and should not be done in the same PR that fixes the linking of static libs.

Reply:
`ssize_t` is already defined before by https://github.com/curl/curl/blob/curl-7_87_0/CMakeLists.txt#L1013-L1026 and https://github.com/curl/curl/blob/curl-7_87_0/lib/curl_config.h.cmake#L776 . It's available for all the sources of curl , but unavailable for `check_symbol_exists()`. I tried to add the same definitions to `CMAKE_REQUIRED_DEFINITIONS` just to make sure `check_symbol_exists(nghttp2_version "nghttp2/nghttp2.h" HAVE_NGHTTP2_DYNAMICLIB)` and `check_symbol_exists(nghttp2_version "nghttp2/nghttp2.h" HAVE_NGHTTP2_STATICLIB)` can work correctly, this setting do not take any effect on curl building.

Reply:
thanks for explaining, I think I understand now. It is necessary for this "hack" to work.

I will leave the judgement for if this is worth merging to someone who better understands and reads cmake.

Reply:
> thanks for explaining, I think I understand now. It is necessary for this "hack" to work.
> 
> I will leave the judgement for if this is worth merging to someone who better understands and reads cmake.

Thanks a lot.